File: FirstAndLastPosition.java
//The approach here is to find the first occurence only first using Binary Search and then use that as the starting point for the Binary Search to find the last occurence.

class Solution {
    private int firstOccurence(int[] nums, int low, int high, int target){
        while(low<=high){
            int mid = low + (high - low)/2;
            if(nums[mid] == target){
                if(mid == 0 || nums[mid] != nums[mid-1]){
                    return mid;
                } else {
                    high = mid - 1;
                }
            }
            else if(nums[low]<=target && nums[mid]>target){
                high = mid -1;
            }
            else {
                low = mid +1;
            }
        }
        return -1;
    }

    private int lastOccurence(int[] nums, int low, int high, int target){
        while(low<=high){
            int mid = low + (high - low)/2;
            if(nums[mid] == target){
                if(mid == nums.length-1 || nums[mid] != nums[mid+1]){
                    return mid;
                } else {
                    low = mid + 1;
                }
            }
            else if(nums[low]<=target && nums[mid]>target){
                high = mid -1;
            }
            else {
                low = mid +1;
            }
        }
        return -1;
    }
    public int[] searchRange(int[] nums, int target) {
        int low = 0;
        int high = nums.length -1;
        if(nums.length == 0) return new int[] {-1,-1};
        int first = firstOccurence(nums, low, high, target);
        if(first == -1) return new int[] {-1,-1};
        int last = lastOccurence(nums, first, high, target);
        return new int[] {first, last};


        
    }
}

File: FirstandLastPosition.py
from typing import List

# Time Complexity : 0(logn)
# Space Complexity : 0(1)
# Did this code successfully run on Leetcode : Yes
# Any problem you faced while coding this : NA

class Solution:
    def searchRange(self, nums: List[int], target: int) -> List[int]:
        if not nums or target > nums[-1] or target < nums[0]:
            return [-1,-1]
        first = self.binarysearchFirst(nums, target)
        if first == -1:
            return [-1,-1]
        second = self.binarysearchSecond(nums,target)
        return [first,second]

    def binarysearchFirst(self, nums,target):
        low = 0 
        high = len(nums) -1

        while low <= high:
            mid = low + (high - low)//2
            if nums[mid] == target:
                if mid == 0 or nums[mid] != nums[mid -1]:
                    return mid
                high = mid -1
            elif nums[mid] < target:
                low = mid +1
            else:
                high = mid -1
        return -1

    def binarysearchSecond(self, nums, target):
        low = 0 
        high = len(nums) -1